### ISSUE TITLE
Commenting/uncommenting code in non-english keyboard layouts?

### ISSUE THREAD

**Issue by martgnz** [observablehq/feedback#7]
**Is your feature request related to a problem? Please describe.**
It's hard to iterate with Observable with a Spanish keyboard layout (I think it must be the same for other variants too e.g. German), because the standard Sublime commenting/uncommenting shortcut (which translates to <kbd>Cmd + Shift + 7</kbd>) [is not supported in Codemirror](https://github.com/codemirror/CodeMirror/issues/6175). Instead of commenting code the browser opens the help menu.

**Describe the solution you'd like**
The best solution would be to support alternative keyboard layouts. But that seems quite hard to do.

A simpler solution would be to add more shortcuts for commenting/uncommenting code (maybe <kbd>Cmd + Shift + 7</kbd> could be added straight away). That'd solve the problem for at least Spanish and German users).

**Comment by jrus** [comment]
One workaround would be to use your operating system to modify your keyboard behavior. There are a wide variety of available keyboard remapping / macro / etc. tools on every platform. For example, on a Mac laptop I have been using [Karabiner-Elements](https://karabiner-elements.pqrs.org) to modify my keyboard behavior.

**Comment by jrus** [comment]
Oh wait, you are saying that **⌘/ does not work for you** as a shortcut for comment/uncomment in Observable cell editors? It does for me. (Mac Safari, using a keyboard that pretends to be US-ANSI layout)

**Comment by jrus** [comment]
In both Chrome and Firefox, with an Spanish ISO keyboard layout, ⌘⇧7 pops up the help menu, presumably because US-ANSI has ? as ⇧/, so that ⌘⇧/ is interpreted as ⌘? which pops up a help menu. In Safari, to get a help menu on an ISO Spanish keyboard, press ⌘⇧' as expected (in ISO Spanish keyboards, ⇧' = ?)

Arguably this is a Chrome/Firefox bug.

But anyway, I expect this will make it impossible to adopt ⌘⇧7 as an in-browser shortcut with a Spanish keyboard on Firefox/Chrome.

**Comment by martgnz** [comment]
@jrus thanks for your thorough investigation, I really appreciate it. I never tried Observable in Safari before: I just took a look and commenting/uncommenting works perfectly with <kbd>Cmd + Shift + 7</kbd>!

I've checked bugzilla and it seems like a long standing issue with non-US layouts ([1](https://bugzilla.mozilla.org/show_bug.cgi?id=311665), [2](https://bugzilla.mozilla.org/show_bug.cgi?id=69230), [3](https://bugzilla.mozilla.org/show_bug.cgi?id=641194), [4](https://bugzilla.mozilla.org/show_bug.cgi?id=670336), [5](https://bugzilla.mozilla.org/show_bug.cgi?id=177508)).

**Comment by severo** [comment]
Same problem with a French keyboard.

I can't use Ctrl+Shift+/ using the "/" key located two keys on the left of the right Shift key (how hard to describe the position of a key!). With @jrus's notebook, I get:

```json
[
  {
    "type": "keydown",
    "key": "Control",
    "code": "ControlLeft",
    "shiftKey": false,
    "metaKey": false,
    "altKey": false,
    "ctrlKey": true,
    "repeat": false,
    "isComposing": false,
    "location": 1,
    "modifierState": ["Control", "NumLock"]
  },
  {
    "type": "keydown",
    "key": "Shift",
    "code": "ShiftLeft",
    "shiftKey": true,
    "metaKey": false,
    "altKey": false,
    "ctrlKey": true,
    "repeat": false,
    "isComposing": false,
    "location": 1,
    "modifierState": ["Control", "NumLock", "Shift"]
  },
  {
    "type": "keydown",
    "key": "/",
    "code": "Period",
    "shiftKey": true,
    "metaKey": false,
    "altKey": false,
    "ctrlKey": true,
    "repeat": false,
    "isComposing": false,
    "location": 0,
    "modifierState": ["Control", "NumLock", "Shift"]
  },
  {
    "type": "keyup",
    "key": "/",
    "code": "Period",
    "shiftKey": true,
    "metaKey": false,
    "altKey": false,
    "ctrlKey": true,
    "repeat": false,
    "isComposing": false,
    "location": 0,
    "modifierState": ["Control", "NumLock", "Shift"]
  },
  {
    "type": "keyup",
    "key": "Shift",
    "code": "ShiftLeft",
    "shiftKey": false,
    "metaKey": false,
    "altKey": false,
    "ctrlKey": true,
    "repeat": false,
    "isComposing": false,
    "location": 1,
    "modifierState": ["Control", "NumLock"]
  },
  {
    "type": "keyup",
    "key": "Control",
    "code": "ControlLeft",
    "shiftKey": false,
    "metaKey": false,
    "altKey": false,
    "ctrlKey": false,
    "repeat": false,
    "isComposing": false,
    "location": 1,
    "modifierState": ["NumLock"]
  }
]
``` 

If I plug an external USB keyboard (also a French keyboard) with a Numpad, I'm able to comment with Ctrl+Shift+/, using the Numpad "/" key. The output is:

```json
[
  {
    "type": "keydown",
    "key": "Control",
    "code": "ControlLeft",
    "shiftKey": false,
    "metaKey": false,
    "altKey": false,
    "ctrlKey": true,
    "repeat": false,
    "isComposing": false,
    "location": 1,
    "modifierState": ["Control", "NumLock"]
  },
  {
    "type": "keydown",
    "key": "Shift",
    "code": "ShiftLeft",
    "shiftKey": true,
    "metaKey": false,
    "altKey": false,
    "ctrlKey": true,
    "repeat": false,
    "isComposing": false,
    "location": 1,
    "modifierState": ["Control", "NumLock", "Shift"]
  },
  {
    "type": "keydown",
    "key": "/",
    "code": "NumpadDivide",
    "shiftKey": true,
    "metaKey": false,
    "altKey": false,
    "ctrlKey": true,
    "repeat": false,
    "isComposing": false,
    "location": 3,
    "modifierState": ["Control", "NumLock", "Shift"]
  },
  {
    "type": "keyup",
    "key": "/",
    "code": "NumpadDivide",
    "shiftKey": true,
    "metaKey": false,
    "altKey": false,
    "ctrlKey": true,
    "repeat": false,
    "isComposing": false,
    "location": 3,
    "modifierState": ["Control", "NumLock", "Shift"]
  },
  {
    "type": "keyup",
    "key": "Shift",
    "code": "ShiftLeft",
    "shiftKey": false,
    "metaKey": false,
    "altKey": false,
    "ctrlKey": true,
    "repeat": false,
    "isComposing": false,
    "location": 1,
    "modifierState": ["Control", "NumLock"]
  },
  {
    "type": "keyup",
    "key": "Control",
    "code": "ControlLeft",
    "shiftKey": false,
    "metaKey": false,
    "altKey": false,
    "ctrlKey": false,
    "repeat": false,
    "isComposing": false,
    "location": 1,
    "modifierState": ["NumLock"]
  }
]
```

**Comment by severo** [comment]
See also: https://talk.observablehq.com/t/ctrl-not-toggling-comments-in-firefox

**Comment by jrus** [comment]
Yeah, it is the same issue. The ISO AZERTY keyboard requires shift to type /, and Mac Firefox/Chrome then misinterpret this as a question mark (due to confusion with US-ANSI). This does not happen with the numpad because shift is not required there to type /. Not sure what happens on non-Mac platforms.

![3LL8a](https://user-images.githubusercontent.com/83411/109904264-d143aa80-7c51-11eb-9047-e3cc0f91eb39.png)

**Comment by severo** [comment]
Thanks. I prefer to avoid tweaking my OS configuration just for that and instead wait for the bug to magically disappear :)

By the way, one great thing with Observable is to be able to develop from any computer. Even if I generally use the same one, it feels awkward to touch the OS in order to be able to use a specific web app.

**Comment by CobusT** [comment]
This seems like something that needs to be fixed in the browsers. Closing this here, but if anyone in this (very helpful!) thread disagrees, please re-open with your thoughts.

**Comment by mootari** [comment]
@CobusT I don't see how that's a realistic expectation. These problems have to be solved on the web application level. Please keep this issue open.
